pylint/config/option_manager_mixin.py

```diff
@@ -305,23 +304,36 @@ def _parse_toml(
         except KeyError:
             return
         for section, values in sections_values.items():
+            section_name = section.upper()
             # TOML has rich types, convert values to
             # strings as ConfigParser expects.
+            if not isinstance(values, dict):
+                # This class is a mixin the add_message come from the pylinter
+                self.add_message(  # type: ignore
+                    "bad-configuration-option-value", line=0, args=(section, values)
+                )
+                continue
             for option, value in values.items():
                 if isinstance(value, bool):
                     values[option] = "yes" if value else "no"
-                elif isinstance(value, (int, float)):
-                    values[option] = str(value)
                 elif isinstance(value, list):
                     values[option] = ",".join(value)
-            parser._sections[section.upper()] = values  # type: ignore
+                else:
+                    values[option] = str(value)
+            for option, value in values.items():
+                try:
+                    parser.set(section_name, option, value=value)
+                except configparser.NoSectionError:
+                    parser.add_section(section_name)
+                    parser.set(section_name, option, value=value)
 
     def load_config_file(self):
         """Dispatch values previously read from a configuration file to each
         options provider)"""
         parser = self.cfgfile_parser
         for section in parser.sections():
-            for option, value in parser.items(section):
+            items = parser.items(section)
+            for option, value in items:
```

Comment:
```suggestion
            for option, value in parser.items(section):
```

This is not too unreadable, right? Or does this create issues?

Reply:
Nice catch, I probably just needed to do that for debugging.